cli/src/lockfiles/javascript.rs

```diff
@@ -52,8 +52,38 @@ impl Parseable for YarnLock {
 
     /// Parses `yarn.lock` files into a vec of packages
     fn parse(&self) -> ParseResult {
-        let (_, entries) = yarn::parse(&self.0).map_err(|_e| "Failed to parse yarn lock file")?;
-        Ok(entries)
+        let yaml_v2: serde_yaml::Value = match serde_yaml::from_str(&self.0) {
+            Ok(yaml) => yaml,
+            Err(_) => {
+                let (_, entries) =
+                    yarn::parse(&self.0).map_err(|_e| "Failed to parse yarn lock file")?;
+                return Ok(entries);
+            }
+        };
+
+        let mapping = yaml_v2.as_mapping().ok_or("Invalid yarn v2 lock file")?;
+        Ok(mapping
+            .iter()
+            .flat_map(|(_k, v)| v.as_mapping())
+            .flat_map(|package| {
+                let name = package
+                    .get(&"resolution".into())
+                    .and_then(|package| package.as_str())
+                    .and_then(|package| package.rsplit_once("@npm:"))
+                    .map(|(name, _version)| name);
+
+                let version = package
+                    .get(&"version".into())
+                    .and_then(|version| version.as_str());
+
+                name.zip(version)
```

Comment:
This is a **very** permissive that will just skip any package that it can't parse.

It's probably fine, but it makes me nervous that there might be some edge case that isn't in our test file that would cause it to skip a package and nobody would notice.

Reply:
That is accurate, I don't have an issue with throwing hard errors on this either, it's probably more appropriate.

Since the `serde_yaml` parser validates v1/2 already, there's no necessity to be lax here, parsing **should** work for every package. So at the very least we should throw a warning?

Reply:
I think a hard error is appropriate here. We want to know if it ever breaks.

There is a top-level `__metadata` key that will need to be ignored. I believe npm disallows package names that start with `_`, so it should be safe to throw out those keys.